File: src/main/java/com/github/njuro/jard/config/security/jwt/JwtAuthenticationFilter.java
package com.github.njuro.jard.config.security.jwt;

import com.github.njuro.jard.common.Constants;
import com.github.njuro.jard.user.UserFacade;
import java.io.IOException;
import java.util.Arrays;
import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import org.springframework.web.filter.OncePerRequestFilter;

/** Filter which handles authentication via JSON Web Token (JWT) cookies. */
@Component
@Slf4j
public class JwtAuthenticationFilter extends OncePerRequestFilter {

  private final JwtTokenProvider tokenProvider;
  private final UserFacade userFacade;

  @Autowired
  public JwtAuthenticationFilter(JwtTokenProvider tokenProvider, @Lazy UserFacade userFacade) {
    this.tokenProvider = tokenProvider;
    this.userFacade = userFacade;
  }

  @Override
  protected void doFilterInternal(
      @NotNull HttpServletRequest request,
      @NotNull HttpServletResponse response,
      @NotNull FilterChain filterChain)
      throws ServletException, IOException {
    try {
      String jwt = getJwtFromRequest(request);

      if (StringUtils.hasText(jwt) && tokenProvider.validateToken(jwt)) {
        String username = tokenProvider.getUsernameFromJWT(jwt);

        UserDetails userDetails = userFacade.loadUserByUsername(username);
        UsernamePasswordAuthenticationToken authentication =
            new UsernamePasswordAuthenticationToken(
                userDetails, null, userDetails.getAuthorities());
        authentication.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));

        SecurityContextHolder.getContext().setAuthentication(authentication);
      }
    } catch (Exception ex) {
      log.error("Could not set user authentication in security context", ex);
    }

    filterChain.doFilter(request, response);
  }

  /**
   * Finds JWT cookie and extracts token from it.
   *
   * @param request incoming HTTP request
   * @return JWT, or {@code null} if no token was found
   */
  private String getJwtFromRequest(HttpServletRequest request) {
    if (request.getCookies() == null) {
      return null;
    }

    return Arrays.stream(request.getCookies())
        .filter(cookie -> cookie.getName().equals(Constants.JWT_COOKIE_NAME))
        .map(Cookie::getValue)
        .findAny()
        .orElse(null);
  }
}

File: src/main/kotlin/com/github/njuro/jard/rewrite/config/security/jwt/JwtAuthenticationFilter.kt
package com.github.njuro.jard.rewrite.config.security.jwt

import com.github.njuro.jard.rewrite.common.JWT_COOKIE_NAME
import com.github.njuro.jard.user.UserFacade
import mu.KLogging
import org.springframework.context.annotation.Lazy
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/** Filter which handles authentication via JSON Web Token (JWT) cookies.  */
@Component
class JwtAuthenticationFilter(
    private val tokenProvider: JwtTokenProvider,
    @Lazy private val userFacade: UserFacade
) : OncePerRequestFilter() {
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        try {
            val jwt = request.getJwtCookie() ?: return
            if (tokenProvider.validateToken(jwt)) {
                val username = tokenProvider.getUsernameFromJWT(jwt)
                val userDetails = userFacade.loadUserByUsername(username)
                val authentication = UsernamePasswordAuthenticationToken(
                    userDetails, null, userDetails.authorities
                )
                authentication.details = WebAuthenticationDetailsSource().buildDetails(request)
                SecurityContextHolder.getContext().authentication = authentication
            }
        } catch (ex: Exception) {
            Companion.logger.error(ex) { "Could not set user authentication in security context" }
        }

        filterChain.doFilter(request, response)
    }

    /**
     * Finds JWT cookie and extracts token from it.
     *
     * @param request incoming HTTP request
     * @return JWT, or `null` if no token was found
     */
    private fun HttpServletRequest.getJwtCookie(): String? = cookies?.find { it.name == JWT_COOKIE_NAME }?.value



    companion object: KLogging()
}
